node_manager.go

```diff
@@ -0,0 +1,992 @@
+	// Set a duration time based on the time-to-live, this will be a
+        // sort of 'duration' time.
+	sleep_duration := time.Duration(lease.TTL / 2)
+
+	// infinite loop
+	for {
+
+		// update the time-to-live
+		inst.updateTTL(lease.ID)
+
+		// then go back to sleep
+		time.Sleep(time.Second * sleep_duration)
+	}
+}
+
+//! Prime a given node to make it ready for jobs.
+/*
+ * @param    chan    channel for listening who is primed
+ *
+ * @return   none
+ */
+func (inst *EtcdInstance) primedLock(primedNotificationChan chan bool) {
+
+	// Grant the lease.
+	lease, err := inst.internal.Grant(context.TODO(), primedTTL)
+
+	// if an error occurs, print it out
+	if err != nil {
+		stdlog("primedLock(): failed to issue a TTL for the future primed lock: " + err.Error())
+		return
+	}
+
+	// Setup a key-value store.
+	kvc := clientv3.NewKV(inst.internal)
+
+	// Grab the background context.
+	ctx, cancel := context.WithTimeout(context.Background(), etcdGracePeriod*time.Second)
+
+	// Check if the value exists, and insert it if it does not.
+	response, err := kvc.Txn(ctx).
+		If(clientv3.Compare(clientv3.Value(primed), "=", "primed")).
+		Then().
+		Else(clientv3.OpPut(primed, "primed", clientv3.WithLease(lease.ID))).
+		Commit()
+
+	// since this is done, cancel the current context
+	cancel()
+
+	// if an error occurred during the, print it out
+	if err != nil {
+		stdlog("primedLock() --> error occured whilst priming lock")
+		stdlog(err.Error())
+		return
+	}
+
+	// if no other node response, this probably means that we are primed
+	// node, so keep the lease entry alive
+	if !response.Succeeded {
+		go inst.keepKeyAlive(lease)
+		primedNotificationChan <- true
+	}
+
+	// if a node *did* response, then this node is not the primed node; go
+	// ahead and clean up the lease ref whilst waiting for a turn...
+	if response.Succeeded {
+		inst.internal.Revoke(context.TODO(), lease.ID)
+		go inst.watchUntilPrimed(primedNotificationChan)
+	}
+
+	// if the node got this far, close the notification channel
+	primedNotificationChan <- false
+}
+
+//! Listens on the notification channel until the current primed node
+//! has completed it's job.
+/*
+ * @param    chan    notification channel
+ *
+ * @return   none
+ */
+func (inst *EtcdInstance) watchUntilPrimed(primedNotificationChan chan bool) {
+
+	// Check for a channel response.
+	responsing_chan := inst.internal.Watch(context.Background(), primed)
+
+	// infinite, keep running until this node gets to be primed
+	for {
+
+		// Grab the response from the channel.
+		wresponse := <-responsing_chan
+
+		// Cycle thru the events.
+		for _, ev := range wresponse.Events {
+
+			// if not a deleted prime event, skip it...
+			if ev.Type != mvccpb.DELETE {
+				continue
+			}
+
+			// otherwise the primed node is complete, so this node needs to
+			// vie for prime lock
+			stdlog("watchUntilPrimed(): main node no longer primed, " +
+				"attempting to prime this node...")
+			inst.primedLock(primedNotificationChan)
```

Comment:
looks like infinite recursion will be here as this will be triggered on itself becoming primed?